vslib/src/sai_vs_hostintf.cpp

```diff
@@ -29,7 +67,414 @@ sai_status_t vs_send_hostif_packet(
+                sai_serialize_object_type(ot).c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    // validate SAI_HOSTIF_ATTR_NAME
+
+    auto attr_name = sai_metadata_get_attr_by_id(SAI_HOSTIF_ATTR_NAME, attr_count, attr_list);
+
+    if (attr_name == NULL)
+    {
+        SWSS_LOG_ERROR("attr SAI_HOSTIF_ATTR_NAME was not passed");
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (strnlen(attr_name->value.chardata, sizeof(attr_name->value.chardata)) >= MAX_INTERFACE_NAME_LEN)
+    {
+        SWSS_LOG_ERROR("interface name is too long: %.*s", MAX_INTERFACE_NAME_LEN, attr_name->value.chardata);
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (strncmp(attr_name->value.chardata, "Ethernet", 8) != 0)
+    {
+        SWSS_LOG_ERROR("interface name should start with EthernetX but is %s", attr_name->value.chardata);
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    std::string name = std::string(attr_name->value.chardata);
+
+    // create TAP device
+
+    SWSS_LOG_INFO("creating hostif %s", name.c_str());
+
+    int tapfd = vs_create_tap_device(name.c_str(), IFF_TAP | IFF_MULTI_QUEUE | IFF_NO_PI);
+
+    if (tapfd < 0)
+    {
+        SWSS_LOG_ERROR("failed to create TAP device for %s", name.c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    SWSS_LOG_INFO("created TAP device for %s, fd: %d", name.c_str(), tapfd);
+
+    // XXX currently tapfd is ignored, it should be closed on hostif_remove
+    // and we should use it to read/write packets from that interface
+
+    sai_attribute_t attr;
+
+    attr.id = SAI_SWITCH_ATTR_SRC_MAC_ADDRESS;
+
+    sai_status_t status = vs_generic_get(SAI_OBJECT_TYPE_SWITCH, switch_id, 1, &attr);
+
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("failed to get SAI_SWITCH_ATTR_SRC_MAC_ADDRESS on switch %s: %s",
+                sai_serialize_object_id(switch_id).c_str(),
+                sai_serialize_status(status).c_str());
+    }
+
+    int err = vs_set_dev_mac_address(name.c_str(), attr.value.mac);
+
+    if (err < 0)
+    {
+        SWSS_LOG_ERROR("failed to set MAC address %s for %s",
+                sai_serialize_mac(attr.value.mac).c_str(),
+                name.c_str());
+
+        close(tapfd);
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    err = ifup(name.c_str());
+
+    if (err < 0)
+    {
+        SWSS_LOG_ERROR("failed to bring ifup %s", name.c_str());
+
+        close(tapfd);
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (!hostif_create_tap_veth_forwarding(name, tapfd))
+    {
+        SWSS_LOG_ERROR("forwarding rule on %s was not added", name.c_str());
+    }
+
+    // TODO what about FDB entries notifications, they also should
+    // be generated if new mac addres will show up on the interface/arp table
+
+    // TODO IP address should be assigned when router interface is created
+
+    SWSS_LOG_INFO("created tap interface %s", name.c_str());
+
+    return vs_generic_create(object_type,
```

Comment:
>return [](start = 4, length = 6)

Close tapfd before teturn?

Reply:
no, tap descriptor is used in thread to write packets